### PR TITLE
fix(dspy): support dspy 2.5 and above

### LINKED ISSUE #1050: DSPy version 2.5 onward logging not compatible with Phoenix
**Is your feature request related to a problem? Please describe.**
Version 2.5 of the DSPy library is not supported by the `openinference-instrumentation-dspy` library.

**Describe the solution you'd like**
Add support for DSPy 2.5 to the `openinference-instrumentation-dspy` library..

**Describe alternatives you've considered**
None.

**Additional context**
DSPy is now using LightLLM for clients and this broke backward compatibly with tracing libraries.

### PR DESCRIPTION
**Pull request by axiomofjoy**
DSPy recently introduced a newly implemented `dspy.LM` interface and is deprecating their old LM interfaces. This PR drops support for those old interfaces. It also switches our tests onto `pytest-recording`.

resolves #1050

### CODE REVIEW

File: python/instrumentation/openinference-instrumentation-dspy/examples/lm.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+from opentelemetry.sdk import trace as trace_sdk
+from opentelemetry.sdk.trace.export import ConsoleSpanExporter, SimpleSpanProcessor
+
+from openinference.instrumentation.dspy import DSPyInstrumentor
+
+endpoint = "http://localhost:6006/v1/traces"
+tracer_provider = trace_sdk.TracerProvider()
+tracer_provider.add_span_processor(SimpleSpanProcessor(OTLPSpanExporter(endpoint)))
+tracer_provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))
+trace_api.set_tracer_provider(tracer_provider=tracer_provider)
```
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-dspy/examples/lm.py:13]:
```suggestion
```

File: python/instrumentation/openinference-instrumentation-dspy/examples/lm.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+DSPyInstrumentor().instrument(tracer_provider=tracer_provider)
+
+
+lm = LM("openai/gpt-4")
```
Comment (RogerHYang) [inline python/instrumentation/openinference-instrumentation-dspy/examples/lm.py:17]:
```suggestion
lm = LM("openai/gpt-4o-mini")
```

**Review by RogerHYang — APPROVED** [review]
does this need to be a major version bump, i.e. `feat!`?

File: python/instrumentation/openinference-instrumentation-dspy/examples/lm.py (hunk 2 above)
Comment (axiomofjoy) [reply python/instrumentation/openinference-instrumentation-dspy/examples/lm.py:17]: Good idea. Will address in a later PR so I don't have to rebase.

File: python/instrumentation/openinference-instrumentation-dspy/examples/lm.py (hunk 1 above)
Comment (axiomofjoy) [reply python/instrumentation/openinference-instrumentation-dspy/examples/lm.py:13]: Good catch.

**Pull request merged**